docs/architecture/adr-025-commit.md

```diff
@@ -60,19 +80,54 @@ const (
 
 ```
 
-Note the need for an extra byte to indicate whether the signature is for the BlockID or for nil.
-This byte can also be used to indicate an absent vote, rather than using a nil object like we currently do,
-which has been [problematic for compatibility between Amino and proto3](https://github.com/tendermint/go-amino/issues/260).
-
-Note we also continue to store the `ValidatorAddress` in the `CommitSig`.
-While this still takes 20-bytes per signature, it ensures that the Commit has all
-information necessary to reconstruct Vote, which simplifies mapping between Commit and Vote objects
-and with debugging. It also may be necessary for the light-client to know which address a signature corresponds to if
-it is trying to verify a current commit with an older validtor set.
+Re the concerns outlined in the context:
+
+**Timestamp**: Leave the timestamp for now. Removing it and switching to
+proposer based time will take more analysis and work, and will be left for a
+future breaking change. In the meantime, the concerns with the current approach to
+BFT time [can be
+mitigated](https://github.com/tendermint/tendermint/issues/2840#issuecomment-529122431).
+
+**ValidatorAddress**: we include it in the `CommitSig` for now. While this
+does increase the block size unecessarily (20-bytes per validator), it has some ergonomic and debugging advantages:
+
+- `Commit` contains everything necessary to reconstruct `[]Vote`, and doesn't depend on additional access to a `ValidatorSet`
+- Lite clients can check if they know the validators in a commit without
+  re-downloading the validator set
+- Easy to see directly in a commit which validators signed what without having
+  to fetch the validator set
+
+If and when we change the CommitSig again, for instance to remove the timestamp,
+we can reconsider whether the ValidatorAddress should be removed.
+
+**Absent Votes**: we include absent votes explicitly with no Signature or
+Timestamp but with the ValidatorAddress. This should resolve the serialization
+issues and make it easy to see which validator's votes failed to be included.
+
+**Other BlockIDs**: We use a single byte to indicate which blockID a `CommitSig`
+is for. The only options are:
+    - `Absent` - no vote received from the this validator, so no signature
+    - `Nil` - validator voted Nil - meaning they did not see a polka in time
+    - `Commit` - validator voted for this block
+
+Note this means we don't allow votes for any other blockIDs. If a signature is
+included in a commit, it is either for nil or the correct blockID. According to
+the Tendermint protocol and assumptions, there is no way for a correct validator to
+precommit for a conflicting blockID in the same round an actual commit was
+created. This was the consensus from
+[#3485](https://github.com/tendermint/tendermint/issues/3485)
+
+We may want to consider supporting other blockIDs later, as a way to capture
```

Comment:
Won't it be possible to capture evidence without including votes for other blockIDs into `CommitSig`? The same question for `nil`. If it's possible, then we won't have to include votes for nil / other block IDs.

Reply:
We want to include nil to signal about validator liveness since its not necessarily a validators fault if they voted nil. But it likely is their fault for voting a different block ID (unless +1/3 are malicious). I'm not sure exactly the case here where it would help to capture this in the commit but I figured I'd just leave the note so that we think about it later. Meanwhile, I think we move forward with including votes for nil but not votes for other blockids.